### ISSUE TITLE
core: Non-compliant YAML spec

### ISSUE THREAD

**Issue by seeyabye** [Jackett/Jackett#11917]
This was something that I noticed while working on https://github.com/Jackett/Jackett/pull/11907 and I'm not quite sure the reasoning behind such implementation and I wanted to seek an answer from the issue. 

I found that under most cardigann based indexers, under `Fields`, one could define repeated keys as follows:

Extracted from `1337x.yml`
```yaml
search:
  fields:
    # dates come in three flavours:
    date:
      # (within this year) 7am Sep. 14th
      optional: true
      selector: td[class^="coll-date"]:not(:contains(":")):not(:contains("'"))
      filters:
        - name: re_replace
          args: ["st|nd|rd|th", ""]
        - name: dateparse
          args: "3pm Jan. 2"
    date:
      # (more than a year ago) Apr. 18th '11
      optional: true
      selector: td[class^="coll-date"]:contains("'")
      filters:
        - name: replace
          args: ["'", ""]
        - name: re_replace
          args: ["st|nd|rd|th", ""]
        - name: dateparse
          args: "Jan. 2 06"
    date:
      # (today) 12:25am
      optional: true
      selector: td[class^="coll-date"]:contains(":")
      filters:
        - name: dateparse
          args: "3:04pm"
```

Notice that `date` was repeated three times, but each of the content of dates are different.
Logically, I can understand that we would like to apply all three date to extract as much information as we can, but the definition itself breaks the YAML specs by having duplicated keys (see [here](https://yaml.org/spec/1.2/spec.html#id2764044). The spec reads:

> Mapping
The content of a mapping node is an unordered set of key: value node pairs, with the restriction that each of the keys is unique. YAML places no further restrictions on the nodes. In particular, keys may be arbitrary nodes, the same node may be used as the value of several key: value pairs, and a mapping could even contain itself as a key or a value (directly or indirectly).

and unfortunately, the implementation we have in Jackett simply breaks the YAML spec.

There are several reasons we should try to conform to the spec, and one of them is to ensure we can include automated ways to check the yaml indexers. Another reason is with more 100 cardigann indexers, it's quite difficult to do mass edit by hand. Tools that are capable of bulk editing requires some form of conformity to the spec.

In conclusion, the `Search.Fields` section may require a rewrite to ensure YAML spec conformity.

**Comment by ilike2burnthing** [comment]
In a situation like that, or:
https://github.com/Jackett/Jackett/blob/6a213c6eabd0264392c48564ff16a9035fd13ca4/src/Jackett.Common/Definitions/skytorrents-to.yml#L70-L80
is there another way to get dates/categories which use different formats or locations?

**Comment by seeyabye** [comment]
It probably shouldn't require much modification on the YAML side.

However, the parser will require some major modifications to complement use cases like the one you showed. I have not looked into it in detail though.

In your example, the fields can be rewritten into a list as follows:

```yaml
  fields:
    - category:
       text: other
    - category|noappend:
       # try category=
       selector: a.label[href*="category="]
       optional: true
    - category|noappend:
       # try type=
       selector: a.label[href*="type="]
       optional: true
```

**Comment by garfield69** [comment]
using the category key as inspiration, almost all the other instances where we have duplicate keys are ones where we used the `optional: true` option in order to handle selectors that may be missing.
the best way to fix this IMO is to introduce `|optional` as a switch in the same way `|noappend`  is used for category.
this switch would replaces the `optional: true` function in the cardigann code.
so in the example provided in the original post, this would become
```yaml
search:
  fields:
    # dates come in three flavours:
    date|optional:
      # (within this year) 7am Sep. 14th
      selector: td[class^="coll-date"]:not(:contains(":")):not(:contains("'"))
      filters:
        - name: re_replace
          args: ["st|nd|rd|th", ""]
        - name: dateparse
          args: "3pm Jan. 2"
    date|optional:
      # (more than a year ago) Apr. 18th '11
      selector: td[class^="coll-date"]:contains("'")
      filters:
        - name: replace
          args: ["'", ""]
        - name: re_replace
          args: ["st|nd|rd|th", ""]
        - name: dateparse
          args: "Jan. 2 06"
    date|optional:
      # (today) 12:25am
      selector: td[class^="coll-date"]:contains(":")
      filters:
        - name: dateparse
          args: "3:04pm"
```
which does the same job and complies with the strict yaml definition.

**Comment by garfield69** [comment]
Hmm. the visual studio is still flagging my `date|optional` as a duplicate key.
why is `category|noappend` not flagged in the same way?

**Comment by seeyabye** [comment]
Well, that's because here, `date|optional` is in a dictionary of `fields` while in the former each `category` is contained within a list/array. Technically, having duplicates of keys in a dictionary is incorrect which is why you're flagged. The `-` makes a difference.

**Comment by garfield69** [comment]
so what you are telling me, if I understand things correctly, is that the only way to be complient is to change the multi date parsing into something like:
```yaml
search:
  fields:
    dateN:
      text: now
    # dates come in three flavours:
    dateW:
      # (within this year) 7am Sep. 14th
      optional: true
      selector: td[class^="coll-date"]:not(:contains(":")):not(:contains("'"))
      filters:
        - name: re_replace
          args: ["st|nd|rd|th", ""]
        - name: dateparse
          args: "3pm Jan. 2"
    dateY:
      # (more than a year ago) Apr. 18th '11
      optional: true
      selector: td[class^="coll-date"]:contains("'")
      filters:
        - name: replace
          args: ["'", ""]
        - name: re_replace
          args: ["st|nd|rd|th", ""]
        - name: dateparse
          args: "Jan. 2 06"
    dateT:
      # (today) 12:25am
      optional: true
      selector: td[class^="coll-date"]:contains(":")
      filters:
        - name: dateparse
          args: "3:04pm"
    date:
      text: "{{ if or .Result.dateW .Result.dateY .Result.dateT .Result.dateN }}{{ or .Result.dateW .Result.dateY .Result.dateT .Result.dateN }}"
```

**Comment by seeyabye** [comment]
There are a couple of ways to do this. One way is as you've suggested, to make each `date` key unique, but it's a problem if we are required to consider the order of execution for each date selector as by specification, these `key: value` pairs are unordered.

The most logical choice would be to reconsider changing `fields` to a sequence. This would allow duplicates and we can write something like the following:

```yaml
search:
  fields:
    - date:
      text: now
    # dates come in three flavours:
    - date:
      # (within this year) 7am Sep. 14th
      optional: true
      selector: td[class^="coll-date"]:not(:contains(":")):not(:contains("'"))
      filters:
        - name: re_replace
          args: ["st|nd|rd|th", ""]
        - name: dateparse
          args: "3pm Jan. 2"
    - date:
      # (more than a year ago) Apr. 18th '11
      optional: true
      selector: td[class^="coll-date"]:contains("'")
      filters:
        - name: replace
          args: ["'", ""]
        - name: re_replace
          args: ["st|nd|rd|th", ""]
        - name: dateparse
          args: "Jan. 2 06"
    - date:
      # (today) 12:25am
      optional: true
      selector: td[class^="coll-date"]:contains(":")
      filters:
        - name: dateparse
          args: "3:04pm"
    - date:
      text: "{{ if or .Result.dateW .Result.dateY .Result.dateT .Result.dateN }}{{ or .Result.dateW .Result.dateY .Result.dateT .Result.dateN }}"

```

Notice that each date is now prepended with a `-`, which assumes that `fields` is a list/array and not a dictionary.
This subtle change looks straightforward, but I've yet to investigate and survey the impact of this change from all the other yaml indexers.

I don't quite recall precisely, but I think it's not a straightforward implementation of the cardigann's parser in this regard.
